Smt/Term.lean

```diff
@@ -61,29 +64,27 @@ def quoteSymbol (s : String) : String :=
   else s
 
 /-- Print given `Term` in SMT-LIBv2 format. -/
-partial def toString : Term → String
-  | literalT l    => l             -- do not quote theory literals
-  | symbolT n     => quoteSymbol n
-  | t@(arrowT ..) =>
-    let ss := arrowToList t
-    "(" ++ String.intercalate " " (ss.init.map toString) ++ ") "
-        ++ ss.getLast!.toString
-  | t@(appT ..) =>
-    let ts := List.reverse <| appToList t
-    "(" ++ String.intercalate " " (ts.map toString) ++ ")"
-  | forallT n s t => s!"(forall (({quoteSymbol n} {s.toString})) {t.toString})"
-  | existsT n s t => s!"(exists (({quoteSymbol n} {s.toString})) {t.toString})"
-  | letT n t b    => s!"(let (({quoteSymbol n} {t.toString})) {b.toString})"
-  where
-    arrowToList : Term → List Term
-      | arrowT d c => d :: arrowToList c
-      | s          => [s]
-    appToList : Term → List Term
-      | appT f t => t :: appToList f
+protected partial def toSexp : Term → Sexp
+  | literalT l    => sexp!{{l}}             -- do not quote theory literals
+  | symbolT n     => sexp!{{quoteSymbol n}}
+  | t@(arrowT ..) => sexp!{(...{List.intersperse sexp!{->} ((arrowToList t).map Term.toSexp)})}
```

Comment:
Do we really want `->`? Previously we would just send a list of `(argTp)` since that's how parameter lists are expressed in SMT-LIB.

Reply:
Hmm... Maybe if we combine `arrowT` and `forallT` (which we should btw). The old printer for the current `arrowT` is only useful in the context of a `declare-fun` command and is not a valid s-expr.

Reply:
Hm, but `->` doesn't appear in the SMT-LIB standard. Is this another custom thing? My impression is that the "arrow type" always looks like `((ArgSort1) (ArgSort2)) (OutSort)`, or `((a ArgSort1) (b ArgSort2)) (OutSort)` with binders. This is different the formula `(forall (x ArgSort1) ..)`, so I'm not sure it makes sense to unify `forallT` and `arrowT`. One is a formula and the other a sort, different things as far as FOL is concerned.

Reply:
Outside of the syntax for `declare-fun`, the syntax for types of higher order symbols is not really defined by the standard. If we take SMT-LIBv3 into consideration, then their syntax uses `->`. Either way, solvers that reject the `->` syntax also reject the old syntax. That's because (AFAIK) only `cvc5` accepts HOL queries and it only accepts the `->` syntax. Regarding `forallT`, I got ahead of myself there. So, let's not waste time discussing it.

btw, I have been breaking FOL and SMT-LIBv2 rules for a while now 🙂. I am doing so mainly to support `cvc5` features and "leanify" the interface for Lean users and developers.